#### **Data Cleaning**

The process of **detecting**, **correcting**, or **removing corrupt** or **inaccurate records** from a dataset is called **data cleaning**.

Its goal is to improve the quality of data, thereby ensuring more **reliable analytics**, **reporting**, and **modeling**.

Data cleaning is a critical step in data preprocessing because the quality of your data directly affects the performance and outcomes of your data analysis or machine learning models.

In a production environment, I mean an environment where data is collected for immediate analysis. Data cleaning starts at the data collection phase.

Before we proceed, we would like to know what makes a dirty data or why do data needs cleaning? Data needs cleaning because of the following reasons.

1. Missing Values
2. Inconsistencies
3. Duplicate Entries
4. Incorrect or out of range values
5. Typographical Errors
6. Noise(irrelevant entries) and outliers
7. Incorrect data types
8. Data merging issues.

Having a good understanding of these issues is very important because they affect how you plan your data cleaning strategies.


**Data cleaning Techniques and Methods**

1. Handling Missing Values


*   Deletion: Remove rows or columns with too many missing values.

*   Imputation: Replace missing values with descriptive statistics like mean and median for numerical data and mode for categorical data.

* Flagging: Create a binaey indicator to showing that the data is missing if its important(like NA)

2. Removing Duplicates
* Identify duplicate records using unique identifiers or by comparing multiple columns.
* Remove duplicates, but ensure you dont lose important data.

3. Standardize Data Format
* Convert datatypes appropriately
* Normalize text data(trim whitespace, convert cases)

4. Correct Inconsistencies
* Map inconsistent representations to a standard format("NG", "nigeria" to "Nigeria")
* use **regex(regualar expression)** or string matching techniquesfor text correction.
5. Dealing With Outliers
* Use statistical methods to detect outliers (Z-scores or IQR method)
* You can decide to remove or cap the outliers(This is based on your domain knowledge though)

6. Data Integration
* Resolve conflicts when merging by handling mismatched keys or duplicate columns.


#### **Lets Practice Data Cleaning With a Dirty Dataset**

##### **A. PDA**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# import library
import pandas as pd
import numpy as np
from datetime import datetime

In [ ]:
# Read in data
hr = pd.read_csv("/content/drive/MyDrive/Publica notes/Machine Learning Notebooks/SectionA-Introduction to Data and Data Analysis/Datasets/hr_data.csv")

In [ ]:
# Inspect dataset
hr.head()

ID             Name     Age                      Email   Join Date  \
0   1    Chinedu Okoro      29    chinedu.okoro@gmail.com  2023-01-15   
1   2        Ngozi Eze      34      ngozi.eze@hotmail.com  15/01/2023   
2   3       Emeka Uche     NaN       emeka.uche@yahoo.com  2022-12-05   
3   4   Adeola Balogun      27          adeola@@gmail.com  2023/02/30   
4   5  Ifeyinwa Okafor  thirty  ifeyinwa.okafor@gmail.com  2023-03-10   

    Salary     Department  
0  $50,000    Engineering  
1    55000   Engineering   
2    60000          Sales  
3  $45,000             HR  
4    48000      Marketing

From this initial view, I could notice that;
1.  The Age column, some of the values are written in words instead of in digits.
2.  The Email column, some of them contains double "@".
3.  The Salary colum, some contains the dollar symbols and comma.
4.  The Joined Date column, The date formating is not consistent.
5.  The Department column, I can notice the use of Abbreviation amongst full names or words.

In [ ]:
hr.tail()

ID               Name Age                          Email   Join Date  \
95   96  Chidera Nwachukwu  78  chidera.nwachukwu@hotmail.com  2022/12/01   
96   97   Ifedayo Olatunji  96     ifedayo.olatunji@yahoo.com  01-12-2022   
97   98      Adenike Ajayi  55        adenike.ajayi@gmail.com  2023-01-01   
98   99   Bolanle Adekunle  36   bolanle.adekunle@hotmail.com  01/01/2023   
99  100  Adebimpe Ogunleye  64    adebimpe.ogunleye@yahoo.com  2023/02/28   

      Salary        Department  
95    102000         Marketing  
96  $150,000        Operations  
97    125000           Finance  
98   $68,000             Legal  
99     88000  Customer Service

From this view,I could notice;
1. From the Age Column, Some age look like outliers. Its worth taking note of.

In [ ]:
# LEts check the dimension
hr.shape

(100, 7)

In [ ]:
# LEts see what our datset is made of
hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID          100 non-null    int64 
 1   Name        100 non-null    object
 2   Age         99 non-null     object
 3   Email       99 non-null     object
 4   Join Date   100 non-null    object
 5   Salary      100 non-null    object
 6   Department  100 non-null    object
dtypes: int64(1), object(6)
memory usage: 5.6+ KB


From the information above, we already have a lead of what do to our data to ensure a consistency.

1. I noticed that Age andEmail have some missing values.
2. Age data type is in object.
3. Salary data type is in object.
4. Join Date does not look pythonic.
5. Also, the Join Data datatypes is not in datetime format
These might not be the only issues with our data, as we explore more, we might still notice some inconsistencies.

##### **B. Data Cleaning**

**What to do based on what we have observed**
1. Check for duplicates
2. Trim extra spaces
3. Standardize data format
4. Handle missing values
5. COnvert data types
6. Validate email and remove inconsistencies
7. Clean inconsistent entries
8. Rename column

**Lets check for duplicates**

In [ ]:
hr.duplicated().sum()

0

**Lets fix the ID first**

In [ ]:
# Wait, lets fix the ID first
hr["ID"] = hr["ID"].apply(lambda x: f"01{int(x):02d}") # I just padded to add extra 0 with max of 2 digits

In [ ]:
# lets check
hr.head()

ID             Name     Age                      Email   Join Date  \
0  0101    Chinedu Okoro      29    chinedu.okoro@gmail.com  2023-01-15   
1  0102        Ngozi Eze      34      ngozi.eze@hotmail.com  15/01/2023   
2  0103       Emeka Uche     NaN       emeka.uche@yahoo.com  2022-12-05   
3  0104   Adeola Balogun      27          adeola@@gmail.com  2023/02/30   
4  0105  Ifeyinwa Okafor  thirty  ifeyinwa.okafor@gmail.com  2023-03-10   

    Salary     Department  
0  $50,000    Engineering  
1    55000   Engineering   
2    60000          Sales  
3  $45,000             HR  
4    48000      Marketing

**Trim Off Extra Spaces**

In [ ]:
# LEts create a list of columns to remove white spaces from
trim_col = ["Name","Email","Department"]

for col in trim_col:
    if col in hr.columns:
        hr[col] = hr[col].astype(str).str.strip()

In [ ]:
#lets check it
hr.head()

ID             Name     Age                      Email   Join Date  \
0  0101    Chinedu Okoro      29    chinedu.okoro@gmail.com  2023-01-15   
1  0102        Ngozi Eze      34      ngozi.eze@hotmail.com  15/01/2023   
2  0103       Emeka Uche     NaN       emeka.uche@yahoo.com  2022-12-05   
3  0104   Adeola Balogun      27          adeola@@gmail.com  2023/02/30   
4  0105  Ifeyinwa Okafor  thirty  ifeyinwa.okafor@gmail.com  2023-03-10   

    Salary   Department  
0  $50,000  Engineering  
1    55000  Engineering  
2    60000        Sales  
3  $45,000           HR  
4    48000    Marketing

Well, it doesnt seem as though anything changed, but its good to be safe, right?

**Standardize Date Joined**

In [ ]:
# We are using pd.to_datetime()

#hr["Join Date"] = pd.to_datetime(hr["Join Date"], errors="coerce", yearfirst=False)

In [ ]:
# LEts check
#hr["Join Date"].info()

In [ ]:
# lets view
#hr.head()

Note that we have "NaT" in our output dataset. If we roll back you will find out that those that are showing "NaT" were formated like this "2023/02/30".
So we would format it properly and comment out this code "`hr["Join Date"] = pd.to_datetime(hr["Join Date"], errors="coerce", yearfirst=False)`"


In [ ]:
# We will be using the .str.replace() method to fix the problem
#hr['Join Date'] = hr['Join Date'].str.replace("/", "-")

In [ ]:
hr.head()

ID             Name     Age                      Email   Join Date  \
0  0101    Chinedu Okoro      29    chinedu.okoro@gmail.com  2023-01-15   
1  0102        Ngozi Eze      34      ngozi.eze@hotmail.com  15/01/2023   
2  0103       Emeka Uche     NaN       emeka.uche@yahoo.com  2022-12-05   
3  0104   Adeola Balogun      27          adeola@@gmail.com  2023/02/30   
4  0105  Ifeyinwa Okafor  thirty  ifeyinwa.okafor@gmail.com  2023-03-10   

    Salary   Department  
0  $50,000  Engineering  
1    55000  Engineering  
2    60000        Sales  
3  $45,000           HR  
4    48000    Marketing

In [ ]:
#hr["Join Date"].head(20)

In [ ]:
# Lets bring this code back to see if we will still have "NaT"

#hr["Join Date"] = pd.to_datetime(hr["Join Date"], errors="coerce", yearfirst=False)
#hr.head()

We  stil have the same issues, lets check for another method.
You would notice that when you scroll through the join date column, the dates are formatted with different time formats...


From observation, we have this types
```c
ISO Format: "%Y-%m-%d" (e.g., "2023-01-15")
Day-Month-Year: "%d-%m-%Y" (e.g., "15-01-2023")
Month-Day-Year: "%m-%d-%Y" (e.g., "11-25-2021")
```
So I think, we should define a python function that will help solve this challenge.
To do this, I will need to import the datetime package.


In [ ]:
# # Defining a datetime parsing function

# def parse_date(date_str):
#     used_formats = ["%Y-%m-%d", "%d-%m-%Y", "%m-%d-%Y", "%Y/%m/%d", "%d/%m/%Y"]
#     for date_format in used_formats:
#         try:
#             dt = datetime.strptime(date_str, date_format)
#             return dt.strftime("%Y-%m-%d")
#         except ValueError:
#             continue
#     return pd.NaT

In [ ]:
# Lets call in the function
#hr['Join Date'] = hr['Join Date'].apply(parse_date)

In [ ]:
#hr["Join Date"].head()


Even after applying the function we still encounter a NaT.

LEts comment out this code `hr['Join Date'] = hr['Join Date'].str.replace("/", "-")` and add the initial format to our function to the if the problem is solved.

In [ ]:
# # I have added more added to the used_formats list

# def parse_date(date_str):
#     used_formats = ["%Y-%m-%d", "%d-%m-%Y", "%m-%d-%Y"]
#     for date_format in used_formats:
#         try:
#             dt = datetime.strptime(date_str, date_format)
#             return dt.strftime("%Y-%m-%d")
#         except ValueError:
#             continue
#     return pd.NaT

# hr['Join Date'] = hr['Join Date'].apply(parse_date)
# hr["Join Date"].head()

We will comment out the above code. It does not solve our challenge. We will refer back to the previous function

In [ ]:

def parse_date(date_str):
    used_formats = ["%Y-%m-%d", "%d-%m-%Y", "%m-%d-%Y", "%Y/%m/%d", "%d/%m/%Y"]
    for date_format in used_formats:
        try:
            dt = datetime.strptime(date_str, date_format)
            return dt.strftime("%Y-%m-%d")
        except ValueError:
            continue
    return pd.NaT

In [ ]:
# Lets call in the function
hr['Join Date'] = hr['Join Date'].apply(parse_date)

In [ ]:
hr["Join Date"].sort_values().tail()


93    2023-11-10
94    2023-11-10
69    2023-11-15
70    2023-11-15
3            NaT
Name: Join Date, dtype: object

From our output we only have one "NaT". I believe we can sort it out manually...with the correct format..

In [ ]:
# 2023/02/30
hr['Join Date'] = hr['Join Date'].str.replace("NaT", "2023-02-30")

In [ ]:
hr["Join Date"].sort_values().tail()


93    2023-11-10
94    2023-11-10
69    2023-11-15
70    2023-11-15
3            NaT
Name: Join Date, dtype: object

In [ ]:
# LEts conver to datetime format
hr["Join Date"] = pd.to_datetime(hr["Join Date"], errors="coerce", yearfirst=False)

In [ ]:
# LEts check if its coverted already
hr["Join Date"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 100 entries, 0 to 99
Series name: Join Date
Non-Null Count  Dtype         
--------------  -----         
99 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 932.0 bytes


In [ ]:
# Lets check for missing value
hr[pd.isna(hr["Join Date"])]

ID            Name Age              Email Join Date   Salary Department
3  0104  Adeola Balogun  27  adeola@@gmail.com       NaT  $45,000         HR

In [ ]:
# lets fill the missing value with its actual date-2023/02/30
hr["Join Date"].fillna(pd.Timestamp("2023-02-28"),inplace=True)

<ipython-input-501-d0d6b5b521c6>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  hr["Join Date"].fillna(pd.Timestamp("2023-02-28"),inplace=True)


Finally, I have figured out the issue, the date is out of range, February is just 28  or 29 days but 30 days was recorded(2023/02/30) so I adjusted it to 28 days instead.

**Fix the column names**


In [ ]:
# LEts add underscore to the Joined Date
hr.rename(columns={"Join Date": "Joined_Date"}, inplace=True)

In [ ]:
hr.columns

Index(['ID', 'Name', 'Age', 'Email', 'Joined_Date', 'Salary', 'Department'], dtype='object')

**Fix the Age column**

In [ ]:
hr["Age"].unique()

array(['29', '34', nan, '27', 'thirty', '45', '38', '50', '42', '31',
       '36', '44', '39', '28', '33', '40', '32', '99', '48', '46', '52',
       '30', '25', '41', '35', '19', '37', '26', '22', '58', '64', '54',
       '65', '56', '47', '68', '55', '72', '62', '66', '102', '201', '18',
       '57', '86', '67', '78', '96'], dtype=object)

In [ ]:
# Lets change "thirty" to "30"
hr["Age"] = hr["Age"].str.replace("thirty", "30")

In [ ]:
# lets check for missing values
hr["Age"].isna().sum()

1

In [ ]:
# Lets convert the age column from object to integer by making Nan explicit
hr["Age"] = pd.to_numeric(hr["Age"], errors="coerce")

In [ ]:
# Lets fill in the missing value with mean age
hr["Age"] = hr["Age"].fillna(round(hr["Age"].mean()))

In [ ]:
# Convert back to integer to take care of
hr["Age"] = hr[ "Age"].astype(int)

In [ ]:
hr["Age"].unique()

array([ 29,  34,  45,  27,  30,  38,  50,  42,  31,  36,  44,  39,  28,
        33,  40,  32,  99,  48,  46,  52,  25,  41,  35,  19,  37,  26,
        22,  58,  64,  54,  65,  56,  47,  68,  55,  72,  62,  66, 102,
       201,  18,  57,  86,  67,  78,  96])

We are done cleaning the Age column

**Fix the Salary Column**

In [ ]:
hr["Salary"].unique()

array(['$50,000', '55000', '60000', '$45,000', '48000', '$55,000',
       '58,000', '62000', '$70,000', '50000', '$65,000', '70000',
       '$60,000', '53000', '$80,000', '$75,000', '68000', '$72,000',
       '54000', '$100,000', '150000', '$85,000', '90000', '$95,000',
       '67000', '$48,000', '$82,000', '$40,000', '75000', '$88,000',
       '$47,000', '82000', '$68,000', '73000', '$58,000', '71000',
       '49000', '$120,000', '56000', '$67,000', '95000', '45000',
       '$90,000', '$62,000', '77000', '79000', '99000', '72000', '61000',
       '$69,000', '80000', '87000', '$110,000', '65000', '$92,000',
       '$83,000', '94000', '105000', '98000', '$76,000', '86000', '64000',
       '93000', '57000', '88000', '76000', '$105,000', '$115,000',
       '$130,000', '102000', '$150,000', '125000'], dtype=object)

In [ ]:
# First, we will remove the dollar symbol
hr["Salary"] = hr["Salary"].str.replace("$", "")

In [ ]:
# Lets remove the commas
hr["Salary"] = hr["Salary"].str.replace(",", "")

In [ ]:
# Lets remove the white spaces
hr["Salary"] = hr["Salary"].str.strip()

In [ ]:
# Lets convert to the Salary datatype from object to integer
hr["Salary"] = hr["Salary"].astype(int)

In [ ]:
hr["Salary"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 100 entries, 0 to 99
Series name: Salary
Non-Null Count  Dtype
--------------  -----
100 non-null    int64
dtypes: int64(1)
memory usage: 932.0 bytes


We are done cleaning the Salary column...

**Fix the Department Column**

In [ ]:
hr["Department"].unique()

array(['Engineering', 'Sales', 'HR', 'Marketing', 'Operations', 'Finance',
       'Research & Development', 'Legal', 'IT', 'Customer Service'],
      dtype=object)

In [ ]:
convert_ab= {"HR": "Human Resources", "IT":"Information Technology"}
# Lets change this abbrevation to full words
hr["Department"] = hr["Department"].replace(convert_ab)

In [ ]:
hr["Department"].unique()

array(['Engineering', 'Sales', 'Human Resources', 'Marketing',
       'Operations', 'Finance', 'Research & Development', 'Legal',
       'Information Technology', 'Customer Service'], dtype=object)

We are done cleaning the Salary column...

**Fix the email column**

In [ ]:
hr["Email"].unique()

array(['chinedu.okoro@gmail.com', 'ngozi.eze@hotmail.com',
       'emeka.uche@yahoo.com', 'adeola@@gmail.com',
       'ifeyinwa.okafor@gmail.com', 'nan', 'funke.ojo@hotmail.com',
       'tope.adeyemi@yahoo.com', 'uche.nwankwo@gmail.com',
       'ifeoma.chukwu@hotmail.com', 'chukwuemeka.obi@yahoo.com',
       'ijeoma.nnamani@gmail.com', 'nnamdi.anozie@hotmail.com',
       'amina.yusuf@yahoo.com', 'musa.bello@gmail.com',
       'fatima.suleiman@hotmail.com', 'olufemi.akinola@yahoo.com',
       'bolaji.balogun@gmail.com', 'tunde.afolabi@hotmail.com',
       'bisi.adekunle@yahoo.com', 'kunle.adegoke@gmail.com',
       'yetunde.adetola@hotmail.com', 'taiwo.olatunji@yahoo.com',
       'femi.opeyemi@gmail.com', 'niyi.faleye@hotmail.com',
       'kemi.ajayi@yahoo.com', 'remi.akinyemi@gmail.com',
       'sola.akintola@hotmail.com', 'wale.balogun@yahoo.com',
       'rita.ibekwe@gmail.com', 'adaeze.nwosu@hotmail.com',
       'efe.eromosele@yahoo.com', 'damilola.ogunleye@gmail.com',
       'olamid

In [ ]:
# Replace "adeola@@gmail.com" with "adeola.balogun@gmail.com"
hr["Email"] = hr["Email"].str.replace("adeola@@gmail.com", "adeola.balogun@gmail.com")

In [ ]:
# Lets validate the email addresses
hr["Email"].nunique()

100

**Fix outliers from Age column**

In [ ]:
# lets check for outliers
hr["Age"].describe()

count    100.000000
mean      44.790000
std       22.795221
min       18.000000
25%       32.000000
50%       39.500000
75%       50.000000
max      201.000000
Name: Age, dtype: float64

From the describe, we see that the mean age is 44.79 while  the median is 39.50.

In [ ]:
hr["Age"].sort_values().tail()

93     86
96     96
20     99
84    102
86    201
Name: Age, dtype: int64

In [ ]:
# Lets sort the outliers using IQR method
#IQR = Quartile3 – Quartile1
#upper_bound = Q3 +1.5*IQR
#lower_bound = Q1 – 1.5*IQR

In [ ]:
# Calculate the IQR(InterQuatileRange)

Q1 = np.percentile(hr["Age"], 25, method='midpoint')
Q3 = np.percentile(hr["Age"], 75, method='midpoint')
IQR = Q3 - Q1
IQR

18.0

In [ ]:
upper_bound = Q3 + 1.5*IQR
lower_bound = Q1 - 1.5*IQR

In [ ]:
print("Upper Bound:", upper_bound)
print("Lower Bound:", lower_bound)
# Note that anything outside upper_bound, that is, 77 and above is an outlier
# Note that anything outside lower_bound, that is, 5 and below is an outlier

Upper Bound: 77.0
Lower Bound: 5.0


In [ ]:
# I think we should cap the outliers instead of removing them

hr["Age"] = np.where(hr["Age"] > upper_bound, upper_bound, hr["Age"])
hr["Age"] = np.where(hr["Age"] < lower_bound, lower_bound, hr["Age"])

In [ ]:
hr["Age"].max()

77.0

In [ ]:
hr["Age"].min()

18.0

I think we are done with our datacleaning, lets save to csv.


In [ ]:
# Lets confirm to be sure that wwe have done a great job
hr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   ID           100 non-null    object        
 1   Name         100 non-null    object        
 2   Age          100 non-null    float64       
 3   Email        100 non-null    object        
 4   Joined_Date  100 non-null    datetime64[ns]
 5   Salary       100 non-null    int64         
 6   Department   100 non-null    object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 5.6+ KB


In [ ]:
hr.head()

ID             Name   Age                      Email Joined_Date  Salary  \
0  0101    Chinedu Okoro  29.0    chinedu.okoro@gmail.com  2023-01-15   50000   
1  0102        Ngozi Eze  34.0      ngozi.eze@hotmail.com  2023-01-15   55000   
2  0103       Emeka Uche  45.0       emeka.uche@yahoo.com  2022-12-05   60000   
3  0104   Adeola Balogun  27.0   adeola.balogun@gmail.com  2023-02-28   45000   
4  0105  Ifeyinwa Okafor  30.0  ifeyinwa.okafor@gmail.com  2023-03-10   48000   

        Department  
0      Engineering  
1      Engineering  
2            Sales  
3  Human Resources  
4        Marketing

In [ ]:
hr.tail()

ID               Name   Age                          Email Joined_Date  \
95   0196  Chidera Nwachukwu  77.0  chidera.nwachukwu@hotmail.com  2022-12-01   
96   0197   Ifedayo Olatunji  77.0     ifedayo.olatunji@yahoo.com  2022-12-01   
97   0198      Adenike Ajayi  55.0        adenike.ajayi@gmail.com  2023-01-01   
98   0199   Bolanle Adekunle  36.0   bolanle.adekunle@hotmail.com  2023-01-01   
99  01100  Adebimpe Ogunleye  64.0    adebimpe.ogunleye@yahoo.com  2023-02-28   

    Salary        Department  
95  102000         Marketing  
96  150000        Operations  
97  125000           Finance  
98   68000             Legal  
99   88000  Customer Service

In [ ]:
# Lets save the final to cvs file
hr.to_csv("/content/drive/MyDrive/Publica notes/Machine Learning Notebooks/SectionA-Introduction to Data and Data Analysis/Datasets/hr_data_cleaned.csv", index=False)